tests/Credentials/AssumeRoleWithWebIdentityCredentialProviderTest.php

```diff
@@ -396,4 +399,168 @@ public function testCanDisableInvalidIdentityTokenRetries()
+        if ($originalRegion !== false) {
+            putenv("AWS_REGION={$originalRegion}");
+        } else {
+            putenv("AWS_REGION");
+        }
+    }
+
+    public function regionPrecedenceProvider(): array
+    {
+        return [
+            'config overrides env' => ['us-west-2', 'eu-west-1', 'us-west-2', false],
+            'env used when no config' => [null, 'eu-west-1', 'eu-west-1', false],
+            'fallback when neither' => [null, null, 'us-east-1', true],
+            'empty string uses fallback' => ['', null, 'us-east-1', true],
+        ];
+    }
+
+    /**
+     * Tests that correct endpoints are called
+     * @dataProvider endpointProvider
+     */
+    public function testEndpointSelection(
+        string $region,
+        string $expectedEndpoint
+    ): void
+    {
+        $tokenFile = tempnam(sys_get_temp_dir(), 'token');
+        file_put_contents($tokenFile, 'test-token-content');
+
+        $config = [
+            'RoleArn' => self::SAMPLE_ROLE_ARN,
+            'WebIdentityTokenFile' => $tokenFile,
+            'region' => $region
+        ];
+
+        $provider = new AssumeRoleWithWebIdentityCredentialProvider($config);
+
+        // Get the client via reflection
+        $reflection = new \ReflectionClass($provider);
+        $clientProperty = $reflection->getProperty('client');
+        $clientProperty->setAccessible(true);
+        $stsClient = $clientProperty->getValue($provider);
+
+        // Set up middleware to capture the endpoint
+        $capturedEndpoint = null;
+        $stsClient->getHandlerList()->appendBuild(
+            Middleware::tap(
+                function ($cmd, $req) use (&$capturedEndpoint) {
+                    $capturedEndpoint = (string) $req->getUri();
+                }
+            )
+        );
+
+        // Mock the STS response
+        $stsClient->getHandlerList()->setHandler(
+            function ($c, $r) {
+                $result = [
+                    'Credentials' => [
+                        'AccessKeyId'     => 'foo',
+                        'SecretAccessKey' => 'bar',
+                        'SessionToken'    => 'baz',
+                        'Expiration'      => DateTimeResult::fromEpoch(time() + 10)
+                    ],
+                    'AssumedRoleUser' => [
+                        'AssumedRoleId' => 'ARXXXXXXXXXXXXXXXXXXX:test_session',
+                        'Arn' => self::SAMPLE_ROLE_ARN . "/test_session"
+                    ]
+                ];
+                return Promise\Create::promiseFor(new Result($result));
+            }
+        );
+
+        // Invoke the provider to trigger the STS call
+        $provider()->wait();
+
+        // Assert the correct endpoint was called
+        $this->assertEquals(
+            $expectedEndpoint,
+            $capturedEndpoint,
+            "Failed asserting endpoint for region: {$region}"
+        );
+
+        // Cleanup
+        unlink($tokenFile);
```

Comment:
I think that also here using a try {} finally {} block is safer in order to clean up the file.